doc/sphinx/proof-engine/tactics.rst

```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+  side of the hypothesis in the conclusion of the goal; the hypothesis is
+  cleared; if the left-hand side of the hypothesis is a variable, it is
+  substituted everywhere in the context and the variable is removed.
+* :n:`<-` - similar to :n:`->`, but replaces the left-hand side of the hypothesis
+  with the right-hand side of the hypothesis.
+* :n:`[= @intropatterns ]` - If the product is over an equality type,
+  applies either :tacn:`injection` or :tacn:`discriminate`
+  instead of :tacn:`destruct`; if :tacn:`injection` is applicable, the patterns
+  :n:`{+, p}` are used on the hypotheses generated by :tacn:`injection`; if the
+  number of patterns is smaller than the number of hypotheses generated, the
+  pattern :n:`?` is used to complete the list.
+
+**need example for [= ... ]**
+
+* :n:`( @simple_intropattern , ... , @simple_intropattern )` - matches
+  a product over an inductive type :g:`I` with a single
+  constructor that itself has at least `n` arguments.  It performs a case
+  analysis over the hypothesis, as :n:`destruct` would, and applies the
+  patterns to the arguments of the constructor of :g:`I`.  Note that
+  that :n:`({+ p})` is equivalent to :n:`[{+ p}]`);
+* :n:`( @simple_intropattern & ... & @simple_intropattern )` - :n:`({+& p})`
+  is a shortcut for matching with
+  :n:`(p,( ... ,( ..., p ) ... ))`.  It expects the hypothesis to be a sequence of
+  right-associative binary inductive constructors such as :g:`conj` or
+  :g:`ex_intro`; for instance, a hypothesis with type
+  :g:`A /\(exists x, B /\ C /\ D)` can be introduced via pattern
+  :n:`(a & x & b & c & d)`;
+
+
+**need example for & ... &**
+
+**A /\\ (exists x, B /\\ C /\\ D) is not valid for (a & x & b & c & d)**
+
+* :n:`*` - introduces one or more quantified variables from the result
+* :n:`**` - introduces one or more quantified variables or hypotheses from the result until there are
+  no more quantified variables or implications (:g:`->`).  :g:`intros **` is equivalent
+  to :g:`intros`.
+
+* :n:`%term` - a pattern :n:`p{+ %term}` first applies :n:`{+ term}`
+  on the hypothesis to be introduced (as in :n:`apply {+, term}`) prior to the
+  application of the introduction pattern :n:`p`;
+
+  the on-the-fly application of lemmas: :n:`p{+ %term}` where :n:`p`
+  itself is not a pattern for on-the-fly application of lemmas (note:
+  syntax is in experimental stage)  **huh?**
+
+**need example, maybe simplify "intros [= -> %IHl]." from List.v?**
+
+**operconstr: coqtop is not accepting this, not sure why--or what the construct means**
+**is it experimental?**
+
+.. note::
+
+   In :n:`intros @intro_pattern_list`, if the last introduction pattern
+   is a disjunctive or conjunctive pattern
+   :n:`[{+| @intro_pattern_list}]`, the completion of :n:`@intro_pattern_list`
+   so that all the arguments of the i-th constructors of the corresponding
+   inductive type are introduced can be controlled with the following option:
```

Comment:
Couldn't tell if this is important or not, or if it could be said more simply.

Reply:
I just noted that this option, as its name suggests only impacts the last intro pattern. In the other cases, it has no effect and does not make `[]` behave like in SSReflect. I'm wondering what was the objective in introducing this option and whether we should remove it (simpler language, less options, less confusion).

Reply:
The flag `Bracketing Last Introduction Pattern` is indeed a compatibility flag. It would probably be less confusion to eventually deprecate it.

Note that ssreflect introduction patterns extended an original mechanism of introduction patterns (in Coq 6.2) which was "procedural". For better compositionality, the introduction patterns evolved in vanilla Coq into a "structured" notion of introduction patterns but ssreflect kept the initial implementation of Coq 6.2. When the pattern is complete, in the sense of giving an introduction pattern for each of the argument of the constructor, there is no difference. There is a difference when it is incomplete. The original 6.2 implementation is such that, e.g., `[ | ] H` is dependent on the inductive structure of the first hypothesis. For instance, if the goal is `forall n:nat, n = 0 -> 0 = n`, the old implementation made that `[ | ] H` was introducing `H : n=0` in the first resulting goal but `H : nat` (the predecessor) in the second goal. On the contrary, the current (vanilla) design has the property that a list `p1 ... pn` of simple introduction patterns applies pointwise to the next `n` first quantified hypotheses or let-ins of the goal.

I don't know what is the best way to explain the question of completion. Maybe can it be said that if `[pat1 ... patn]` has less simple patterns than the number of arguments in the constructor, it automatically expands to `[pat1 ... patn ? ... ?]` with as many extra introduction patterns so as to match the arity of the constructor. Note that let-ins are included in the count.

Reply:
Yes, I think the explanation you give in the last paragraph is a good one.
@jfehrle Could you deprecate the `Bracketing Last Introduction Pattern` as part of this PR?